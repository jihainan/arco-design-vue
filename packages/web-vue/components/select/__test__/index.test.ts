import { mount } from '@vue/test-utils';
import Select from '../index';

describe('Select', () => {
  test('show dropdown correctly', async () => {
    const wrapper = mount(Select, {
      props: {
        options: ['Beijing', 'Shanghai', 'Guangzhou', 'Chengdu'],
      },
    });

    await wrapper.find('.arco-select-view').trigger('click');

    expect(document.body.outerHTML).toMatchSnapshot();
  });

  test('keyboard correctly', async () => {
    const wrapper = mount(Select, {
      props: {
        options: ['Beijing', 'Shanghai', 'Guangzhou'],
      },
    });

    const baseSelect = wrapper.findComponent({ name: 'BaseSelect' });

    const input = wrapper.find('input');
    await input.trigger('click');
    const dropdown = wrapper.findComponent({ name: 'DropdownPanel' });
    await input.trigger('keydown', { code: 'ArrowUp' });

    expect(dropdown.find('.arco-dropdown-option-active').text()).toBe(
      'Guangzhou'
    );

    await input.trigger('keydown', { code: 'ArrowDown' });

    expect(dropdown.find('.arco-dropdown-option-active').text()).toBe(
      'Beijing'
    );

    await input.trigger('keydown', { code: 'Enter' });

    expect(baseSelect.emitted('change')?.[0]).toEqual(['Beijing']);
  });

  test('should support mouse', async () => {
    const wrapper = mount(Select, {
      props: {
        options: ['Beijing', 'Shanghai', 'Guangzhou'],
      },
    });

    const baseSelect = wrapper.findComponent({ name: 'BaseSelect' });

    const input = wrapper.find('input');
    await input.trigger('click');
    const dropdown = wrapper.findComponent({ name: 'DropdownPanel' });
    const options = dropdown.findAll('.arco-dropdown-option');
    await options[0].trigger('mouseenter');
    await options[0].trigger('mouseleave');
    await options[1].trigger('mouseenter');
    await options[1].trigger('click');

    expect(baseSelect.emitted('change')?.[0]).toEqual(['Shanghai']);
  });

  test('should show search option', async () => {
    const wrapper = mount(Select, {
      props: {
        options: ['Beijing', 'Shanghai', 'Guangzhou'],
        allowSearch: true,
      },
    });

    const input = wrapper.find('input');
    await input.trigger('click');
    await input.setValue('sh');
    const dropdown = wrapper.findComponent({ name: 'DropdownPanel' });
    const option = dropdown.find('.arco-dropdown-option');
    expect(option.text()).toBe('Shanghai');
  });

  test('should enable create option', async () => {
    const wrapper = mount(Select, {
      props: {
        options: ['Beijing', 'Shanghai', 'Guangzhou'],
        allowCreate: true,
      },
    });

    const baseSelect = wrapper.findComponent({ name: 'BaseSelect' });

    const input = wrapper.find('input');
    await input.trigger('click');
    await input.setValue('Xian');
    await input.trigger('keydown', { code: 'Enter' });

    expect(baseSelect.emitted('change')?.[0]).toEqual(['Xian']);
  });
});
