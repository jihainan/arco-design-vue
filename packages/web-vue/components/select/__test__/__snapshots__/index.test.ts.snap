// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select show dropdown correctly 1`] = `"<body><transition-stub><div class=\\"arco-trigger-popup arco-trigger-position-bl\\" style=\\"z-index: 1000; left: 0px; top: 4px; transform-origin: 0 0; width: 0px;\\" trigger-placement=\\"bl\\"><div class=\\"arco-trigger-content\\"><div class=\\"arco-dropdown arco-select-dropdown\\"><div class=\\"arco-spin\\" style=\\"width: 100%; display: block;\\"><!--v-if--><div class=\\"arco-dropdown-list-wrapper\\"><ul class=\\"arco-dropdown-list\\"><li class=\\"arco-dropdown-option arco-dropdown-option-active\\"><!--v-if-->Beijing<!--v-if--></li><li class=\\"arco-dropdown-option\\"><!--v-if-->Shanghai<!--v-if--></li><li class=\\"arco-dropdown-option\\"><!--v-if-->Guangzhou<!--v-if--></li><li class=\\"arco-dropdown-option\\"><!--v-if-->Chengdu<!--v-if--></li></ul></div><!--v-if--><!----></div></div></div><!----></div></transition-stub></body>"`;
